# G.G, Good Game! Investigating the Structural Characteristics that Make Highly-Rated Video Games

# Notebook 1: Loading and Cleaning Data
**Steps to load, clean, and analyze the data upon first inspection.**

&nbsp;

**By: [Lae Rivera](https://github.com/laerivera)**

**BrainStation Data Science**

**April 4, 2022**

In this portion of the notebook, I will first define the purpose of this project where I discuss the reasons for doing my analysis.  I then will begin the fundamental steps to loading and cleaning the data.

***

## Notebook Contents

1. [**Introduction**](#a1)<br>
    1.1 [Problem Statement](#a1.1)<br>
    1.2 [Data Collection](#a1.2)<br>

2. [**Loading & Checking**](#a2)<br>
    2.1 [Data and Column Descriptions](#a2.1)<br>
    
3. [**Investigate Columns**](#a3)<br>
    3.1.[Correcting Data Types](#a3.1)<br>
    3.2.[Remove Null Values](#a3.2)<br>
        
4. [**Creating Features**](#a4)<br>

***

# Introduction <a id="a1"></a>



**Problem Statement** <a id="a1.1"></a>

Through emerging technologies and a fresh load of new releases, the video game industry is expected to be worth over $314 billion USD within the next 4 years. The problem is, recent trends show the highest rated games have been struck with delays, incomplete releases, and poor reviews and ratings—which continues to not only disappoint the community, but lose trust in the quality these releases provide. However, there seems to be games that constantly remain popular, even years after its release, which leaves some to wonder— what do these games have in common? 

The goal of this project is to explore the structural characteristics  that make superior, highly rated video games.  I hope the conclusions drawn by the end of this investigation sheds useful and actionable insights for small, new, or at-home indie game developers on what qualities to include if a game is to receive a high rating score. This would especially help gain some traction to perhaps receive more funding,  better marketing, and simply,  just be recognized.

**Data Collection** <a id="a1.2"></a>

The data used in this project was web-scraped  from the Metacritic website. The dataset was then published on Kaggle by user DeepContractor titled, [Top Video Games 1995-2021 Metacritic](https://www.kaggle.com/datasets/deepcontractor/top-video-games-19952021-metacritic), and contains a large list of video game review scores along with the following attributes:  game title, release date, descriptive game summary/premise, meta score achieved, and user review given. The dataset is updated on an annual basis and contains a mixture of both numeric and non-numeric data types. 

An instant direct download link to the dataset can be found [here](https://drive.google.com/uc?export=download&id=1Se0Uazu9SQn_ifH57VnSDVl1SHnxEFBG)

***

**Importing Libraries**

These will be the libraries I will be needing to conduct analysis on the data.

In [1]:
# Importing libraries

# Main Libraries
import numpy as np
import pandas as pd
import joblib

***

# Loading & Checking Data <a id="a2"></a>

### Data and Column Descriptions <a id="a2.1"></a>

Before any data analysis can begin, I want to take a quick look and explrore the data.

In [2]:
# Read in data from my data folder
games_df = pd.read_csv("data/all_games.csv")

In [3]:
# View first 5 rows
games_df.head() 

name        platform        release_date  \
0  The Legend of Zelda: Ocarina of Time     Nintendo 64   November 23, 1998   
1              Tony Hawk's Pro Skater 2     PlayStation  September 20, 2000   
2                   Grand Theft Auto IV   PlayStation 3      April 29, 2008   
3                           SoulCalibur       Dreamcast   September 8, 1999   
4                   Grand Theft Auto IV        Xbox 360      April 29, 2008   

                                             summary  meta_score user_review  
0  As a young boy, Link is tricked by Ganondorf, ...          99         9.1  
1  As most major publishers' development efforts ...          98         7.4  
2  [Metacritic's 2008 PS3 Game of the Year; Also ...          98         7.7  
3  This is a tale of souls and swords, transcendi...          98         8.4  
4  [Metacritic's 2008 Xbox 360 Game of the Year; ...          98         7.9

In [4]:
# Returns shape of the dataframe
print(f"Shape of the dataset: {games_df.shape}")

Shape of the dataset: (18800, 6)


In [5]:
# Returns the list of columns
print("List of Columns:")
list(games_df.columns)

List of Columns:


['name', 'platform', 'release_date', 'summary', 'meta_score', 'user_review']

***

**Column Description**


`name`: contains the name/title of the video game

`platform`: contains the computer system the game runs on

`release_date`: the date when the game was released; includes the Month, Day and Year (in that order)

`summary`:  the game's premise; includes 'Game of the Year' tag

`meta_score`: meta score received; integer from 0-100

`user_review`: user score received; decimal from 0-10

&nbsp;
***

# Investigate Columns <a id="a3"></a>

**Dataset Summary Information**

In [6]:
# Checking data types and columns
games_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18800 entries, 0 to 18799
Data columns (total 6 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   name          18800 non-null  object
 1   platform      18800 non-null  object
 2   release_date  18800 non-null  object
 3   summary       18686 non-null  object
 4   meta_score    18800 non-null  int64 
 5   user_review   18800 non-null  object
dtypes: int64(1), object(5)
memory usage: 881.4+ KB


In [7]:
#Initial percentage of null values 
print("Initial Percentage of Null Values %")
round(games_df.isna().sum()/len(games_df)*100, 2) #coverted into a percent(%) and round to 2 decimal places

Initial Percentage of Null Values %


name            0.00
platform        0.00
release_date    0.00
summary         0.61
meta_score      0.00
user_review     0.00
dtype: float64

&nbsp;

There seems to be some discrepanices between some columns and its datatype. For example, looking at the dataframe initially, the `user_review` column is *clearly a decimal value*, thus I would expect to see a *float* data type——instead, we can see that it is a string data type. I will investigate the cause of this and make some changes during processing later on. Additionally, there seems to be some null values present under the summary column. This values make up less than ~1% of the dataset and since this amount is so small, I may just drop these rows. However, I will first continue to investigate other columns before dropping anything.

***

## Correcting Data Types <a id="a3.1"></a>

**Change `release_date` to a datetime format**

In [8]:
# Change release date into datetime format
games_df['release_date'] = pd.to_datetime(games_df['release_date']) #overwrite datatype changes to existing column

In [9]:
# Verify Changes
games_df['release_date'].dtypes 

dtype('<M8[ns]')

'*<M8[ns]*' indicates a datetime format!

---

**Change `user_review` to a float/integer**

When attempting to change 'user_review', I noticed **a simple `astype()` code is not working**. Upon investigation, I discovered this was because when the dataset was scraped from the Metacritic website, some newly added games that were listed on the website had **not yet received any User Review scores**, thus leaving a 'to be determined' (tbd) status. (see below)

In [10]:
# Checks the values in 'user_review'
games_df.user_review.unique()

array(['9.1', '7.4', '7.7', '8.4', '7.9', '8.0', '8.3', '6.2', '8.7',
       '7.5', '8.8', '8.5', '8.6', '8.9', '8.2', '9.2', '7.8', '9.0',
       '9.3', '8.1', '6.8', '7.6', '6.9', '6.6', '7.3', '7.2', '9.4',
       '5.7', '6.0', '5.8', '7.1', '3.3', '5.0', '5.5', '6.5', '5.9',
       '6.7', '3.8', '6.4', '7.0', '6.3', '4.9', 'tbd', '5.3', '3.5',
       '4.7', '6.1', '4.8', '4.1', '5.6', '4.4', '5.4', '4.6', '3.1',
       '5.2', '4.3', '1.8', '4.5', '5.1', '4.2', '4.0', '9.7', '3.2',
       '2.9', '2.1', '2.4', '3.4', '3.0', '3.7', '2.2', '2.6', '2.7',
       '2.5', '2.8', '3.9', '2.0', '9.6', '1.5', '3.6', '1.4', '1.2',
       '0.9', '1.9', '1.1', '1.7', '0.8', '1.0', '0.6', '2.3', '0.5',
       '1.6', '0.2', '0.7', '1.3'], dtype=object)

In [11]:
# Relative distribution of unique values in user_review
user_review_percentage = games_df['user_review'].value_counts(normalize=True).sort_index(ascending=False).sort_values(ascending=False)*100.00

#What percentage of whole column this makes up
print(f'Percentage of rows contains `tba` user review: {round(user_review_percentage[0], 2)}%')

Percentage of rows contains `tba` user review: 7.26%


**The amount of 'tbd' rows make up around ~7% of the dataset**. This is far too many rows to drop right away. Since I would like to preserve as much data as possible, I will fill in these missing values. 

---

### Fill in missing values in user_review

**Goal**: 
My goal here is to fill in as much of the 'tbd' user_score with other values (such as an aggregated mean). I want to reduce the amount of missing values to a threshold that is acceptable where I can just drop the rows. In this case, I am aiming for roughly 5% total percentage of nulls before dropping.

**Filling in Plan:**
- Create a dataframe *without* 'tbd' user review scores 
    
- Group this newly created dataframe with only non-tbd values by game `name` and calculate **average `user_review` per game**. This is done because there are instances of rows having the *same* game name, but on different platforms-that have different user scores. Getting an average of these user scores will be a good value to impute to the missing values.
    
- We now have a dataframe of **`game name and average user_review`** to help fill in the missing values

#### Step 1: Isolate non-tbd values

In [12]:
# Create dataframe dropping all user_review with 'tbd' rows to get a clean dataset
games_no_tbd = games_df.drop(games_df[games_df.user_review == 'tbd'].index)

In [13]:
# Turn datatype from object to float so calculations can be done
games_no_tbd['user_review'] = games_no_tbd['user_review'].astype(float)

In [14]:
# Summary statistics for games_no_tbd
games_no_tbd['user_review'].describe()

count    17435.000000
mean         6.990846
std          1.351554
min          0.200000
25%          6.300000
50%          7.300000
75%          7.900000
max          9.700000
Name: user_review, dtype: float64

With non-'tbd' data isolated, I can change the data type of the `user_review` column to a float. This would allow me to apply calculations to this column.

#### Step 2: Calculate average user_review score per game on non-tbd values

In [15]:
# Group by game name and get the average user score
games_no_tbd_avg_user_review = games_no_tbd.groupby(['name'], as_index = False)['user_review'].mean()

In [16]:
# Change column name to reflect mean calculations
games_no_tbd_avg_user_review = games_no_tbd_avg_user_review.rename(columns={'user_review': 'average_user_review'})

In [17]:
games_no_tbd_avg_user_review.head()

name  average_user_review
0                       #IDARB                 6.90
1                'Splosion Man                 7.80
2                     .detuned                 3.70
3      .hack//G.U. Last Recode                 8.15
4  .hack//G.U. vol. 1//Rebirth                 8.10

As seen in the games_no_tbd_avg_user_review dataframe, we grouped all the unique game titles (names) together and aggregated an average user_review score. This dataframe will serve as our reference in imputing the missing values.

#### Step 3: Correct 'tbd' to proper NaN value

Back to the original dataframe, we can replace all 'tbd' values into a proper null value to reflect more clearly the nature of the 'tbd' value—that it is a missing value. This also allows us to calculate all null values using a single code. 

*Note: Pandas treats NaN and Null values the [same](https://www.geeksforgeeks.org/working-with-missing-data-in-pandas/)—both indicative of missing values*

In [18]:
# Replace the 'tbd' with None to make it a true null value
games_df["user_review"].replace('tbd', np.NaN, inplace=True)

Finally, we can change the data type of user_review to a float so it can reflect the true data type of user reviews (decimal number from 0-10).

In [19]:
# Make datatype a float
games_df['user_review'] = games_df["user_review"].astype(float)

In [20]:
# Check datatype
games_df.dtypes

name                    object
platform                object
release_date    datetime64[ns]
summary                 object
meta_score               int64
user_review            float64
dtype: object

`user_review` is now a float data type! 

In [21]:
# Null value count
games_df.isna().sum()

name               0
platform           0
release_date       0
summary          114
meta_score         0
user_review     1365
dtype: int64

Here, we can see that 'tbd' is now a *proper* null value. Earlier we counted the amount of 'tbd' columns, which is 1365. Since I replaced 'tbd' with 'NaN', *these numbers should correspond to each other*. Here, we see the same number of NaN values- showing us the code worked!

#### Step 4: Fill in NaN values with average user_review by matching game names

In [22]:
# Filling matching game names 
games_df.set_index("name", inplace=True)
games_df["user_review"].fillna(games_no_tbd_avg_user_review.set_index("name")["average_user_review"], inplace=True)
games_df.reset_index(inplace=True)

The logic of the code above essentially sets the "name" column as an index to be matched. Next, I am filling in the user_review null values with the "average_user_review" *if* the game names are a match (and if user_review is null). 

In [23]:
# Updated null count 
print("Count of Null Values")
games_df.isna().sum() 

Count of Null Values


name              0
platform          0
release_date      0
summary         114
meta_score        0
user_review     928
dtype: int64

In [24]:
# Updated Percentage of null values 
print("Percentage of Null Values %")
round(games_df.isna().sum()/len(games_df)*100, 2)

Percentage of Null Values %


name            0.00
platform        0.00
release_date    0.00
summary         0.61
meta_score      0.00
user_review     4.94
dtype: float64

After imputing values, the percentage of null values in user_review went from 7.26% to 4.94%! This is a more acceptable drop. 

***

#### Updated Data Types

In [25]:
# Updated datatypes of columns
games_df.dtypes

name                    object
platform                object
release_date    datetime64[ns]
summary                 object
meta_score               int64
user_review            float64
dtype: object

Now the data types are consistent with the information within the columns!

---

## Remove Null Values <a id="a3.2"></a>

In [26]:
print("Percentage of Null Values %")
round(games_df.isna().sum()/len(games_df)*100.00, 2)

Percentage of Null Values %


name            0.00
platform        0.00
release_date    0.00
summary         0.61
meta_score      0.00
user_review     4.94
dtype: float64

The null values in `summary` make up less than 1% of the dataset. This value is not high enough to have an effect on the data should I choose to drop these rows. In total, the number of null values make up a *little over 5% of the data* (this also does not include rows that have both summary and user_review as 'null'). Since this value is low, I can acceptably drop the rows.

In [27]:
# Drop rows that are null
games_df = games_df.dropna(axis=0) 

In [28]:
# Confirm no null values
games_df.isna().sum() 

name            0
platform        0
release_date    0
summary         0
meta_score      0
user_review     0
dtype: int64

**No null values!**

In [29]:
# Reset the index
games_df.reset_index(drop=True, inplace=True)

In [30]:
games_df.head()

name        platform release_date  \
0  The Legend of Zelda: Ocarina of Time     Nintendo 64   1998-11-23   
1              Tony Hawk's Pro Skater 2     PlayStation   2000-09-20   
2                   Grand Theft Auto IV   PlayStation 3   2008-04-29   
3                           SoulCalibur       Dreamcast   1999-09-08   
4                   Grand Theft Auto IV        Xbox 360   2008-04-29   

                                             summary  meta_score  user_review  
0  As a young boy, Link is tricked by Ganondorf, ...          99          9.1  
1  As most major publishers' development efforts ...          98          7.4  
2  [Metacritic's 2008 PS3 Game of the Year; Also ...          98          7.7  
3  This is a tale of souls and swords, transcendi...          98          8.4  
4  [Metacritic's 2008 Xbox 360 Game of the Year; ...          98          7.9

Now there are no null values in the dataset. I also reset the index since I operated on the dataset. This leaves me with a cleaner dataframe that I can continue to work on.

***

# Creating Features <a id="a4"></a>

#### Create a 'Year' and 'Month' column

I now wish to extract the year and the month column so it will be treated as numerical data that can be analyzed more closely. 

**Workflow Plan**:
- I want to extract `year` and `month` details from the `release_date` column—which is a datetime data type. Extracting this will use dt [functions](https://docs.python.org/3/library/datetime.html) such as `dt.year` and `dt.month` to extract the year and month respectively.
- A 'day' column will *not* be created since it will be too trivial a detail to include.
- Year and Month is more than enough information to do analysis with.

In [31]:
# Extracting year and month from date and creating new columns
games_df['release_year'], games_df['release_month'] = games_df['release_date'].dt.year, games_df['release_date'].dt.month


In [32]:
# Dropping original release_date column to avoid multicolliniearity and because it is redundant
games_df = games_df.drop(columns=['release_date'])

In [33]:
# View newly created columns
games_df[['release_year','release_month']].head()

release_year  release_month
0          1998             11
1          2000              9
2          2008              4
3          1999              9
4          2008              4

In [34]:
# View Dataframe after operations
games_df.head() #view first 5 columns

name        platform  \
0  The Legend of Zelda: Ocarina of Time     Nintendo 64   
1              Tony Hawk's Pro Skater 2     PlayStation   
2                   Grand Theft Auto IV   PlayStation 3   
3                           SoulCalibur       Dreamcast   
4                   Grand Theft Auto IV        Xbox 360   

                                             summary  meta_score  user_review  \
0  As a young boy, Link is tricked by Ganondorf, ...          99          9.1   
1  As most major publishers' development efforts ...          98          7.4   
2  [Metacritic's 2008 PS3 Game of the Year; Also ...          98          7.7   
3  This is a tale of souls and swords, transcendi...          98          8.4   
4  [Metacritic's 2008 Xbox 360 Game of the Year; ...          98          7.9   

   release_year  release_month  
0          1998             11  
1          2000              9  
2          2008              4  
3          1999              9  
4          2008              4

As we can see, the dataframe now have `release_year` and `release_month` columns. The dataframe is looking good. I will now save my work into a pickle file and continue my data exploration.

***

**Saving Work**

In [35]:
# Save data as pickle file in my data folder
joblib.dump(games_df, 'data/games_df.pkl')

['data/games_df.pkl']

***

**Notebook Ending Remarks**

In this notebook, I have completed some cleaning steps to ensure the data is of its highest quality before processing may begin. This will give the most accurate analysis, and thus more accurate findings and insights.

In the next notebook, I will be exploring my data in more detail by doing some exploratory data analysis. See Notebook *2-EDA*.

***